backend/qemu.pm

```diff
@@ -346,6 +347,59 @@ sub deflate_balloon ($self) {
     $self->handle_qmp_command({execute => 'balloon', arguments => {value => $ram_bytes}}, fatal => 1);
 }
 
+sub save_storage ($self, $args) {
+    my $vars = \%bmwqemu::vars;
+    my $bdc = $self->{proc}->blockdev_conf;
+    my $fname = $args->{filename};
+    my $rsp = $self->handle_qmp_command({execute => 'query-status'}, fatal => 1);
+    bmwqemu::diag("Saving saving storage (Current VM state is $rsp->{return}->{status})");
+    my $was_running = $rsp->{return}->{status} eq 'running';
+
+    if ($was_running) {
+        $self->inflate_balloon();
+        $self->freeze_vm();
+    }
+
+    mkpath("assets_public");
+    $bdc->for_each_drive(sub ($drive) {
+            my $size = $drive->{drive}->{size};
+            my $id = $drive->{id} . "-backup-$fname";
+            my $node = $drive->{drive}->{node_name};
+            my $my_node = "$node-$fname";
+            my $bck_file = "assets_public/$my_node-$vars->{NAME}.qcow2";
+            # create disk
+            runcmd('qemu-img', 'create', '-f', 'qcow2', "$bck_file", "$size");
+            my $req = {execute => 'blockdev-add',
+                arguments => {driver => 'qcow2', 'node-name' => $my_node,
+                    file => {driver => 'file', filename => $bck_file}
+                }};
+            $self->handle_qmp_command($req, fatal => 1);
+            $req = {execute => 'blockdev-backup',
+                arguments => {device => $node, target => $my_node,
+                    sync => 'full', 'job-id' => $id}};
+            $self->handle_qmp_command($req, fatal => 1);
+            my $return;
+            my $timeout = $vars->{BACKUP_DRIVER_TIMEOUT} // 60 * 15;
+            # backup is backround job -> wait for finish before continue
```

Comment:
background?

Reply:
yes, `blockdev-backup` creates background job ... we can optimise this in case of more disks by runing wait loop outside this loop and backup devices in one go. Question is what happens with IO in that case.

Reply:
To be honest I wouldn't try too hard to optimize it for now. Let's see that it works and get some feedback from real usage.